Empty issuer DN when using selfSigned

**Describe the bug**:

The `selfSigned` issuer seems to generate a certificate that, while tolerated by Chromium for example, is rejected as malformed by Java—apparently due to lack of an issuer DN (and perhaps other reasons if it were to get past that).

**Expected behaviour**:

That the certificate be accepted as well-fomed and importable into the Java cert store.

**Steps to reproduce the bug**:

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: self-signed
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: xxx
spec:
  secretName: xxx-tls
  dnsNames:
  - what.ever
  issuerRef:
    name: self-signed
```

Once the secret is created:

```console
$ kubectl get secret xxx-tls -o go-template='{{ index .data "ca.crt" | base64decode }}' | tee xxx.crt
-----BEGIN CERTIFICATE-----
…
-----END CERTIFICATE-----
$ openssl x509 -text -in xxx.crt
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            …
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: 
        Validity
…
$ keytool -cacerts -storepass changeit -import -file xxx.crt
keytool error: java.lang.Exception: Input not an X.509 certificate
$ keytool -v -printcert -file xxx.crt
keytool error: java.lang.Exception: Failed to parse input
java.lang.Exception: Failed to parse input
	at java.base/sun.security.tools.keytool.Main.printCertFromStream(Main.java:2626)
	at java.base/sun.security.tools.keytool.Main.doPrintCert(Main.java:2780)
	at java.base/sun.security.tools.keytool.Main.doCommands(Main.java:1246)
	at java.base/sun.security.tools.keytool.Main.run(Main.java:405)
	at java.base/sun.security.tools.keytool.Main.main(Main.java:398)
Caused by: java.security.cert.CertificateParsingException: Empty issuer DN not allowed in X509Certificates
	at java.base/sun.security.x509.X509CertInfo.parse(X509CertInfo.java:658)
	at java.base/sun.security.x509.X509CertInfo.<init>(X509CertInfo.java:169)
	at java.base/sun.security.x509.X509CertImpl.parse(X509CertImpl.java:1842)
	at java.base/sun.security.x509.X509CertImpl.<init>(X509CertImpl.java:194)
	at java.base/sun.security.provider.X509Factory.parseX509orPKCS7Cert(X509Factory.java:476)
	at java.base/sun.security.provider.X509Factory.engineGenerateCertificates(X509Factory.java:361)
	at java.base/java.security.cert.CertificateFactory.generateCertificates(CertificateFactory.java:478)
	at java.base/sun.security.tools.keytool.Main.printCertFromStream(Main.java:2624)
	... 4 more
```

**Anything else we need to know?**:

**Environment details:**:
- Kubernetes version: v1.18.14-gke.1600
- Cloud-provider/provisioner: GKE
- cert-manager version: v1.1.0
- Install method: Helm

Java:

```
openjdk version "11.0.9.1" 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```

/kind bug

--- a/pkg/controller/certificaterequests/selfsigned/BUILD.bazel
+++ b/pkg/controller/certificaterequests/selfsigned/BUILD.bazel
@@ -16,8 +16,10 @@ go_library(
         "//pkg/util/errors:go_default_library",
         "//pkg/util/kube:go_default_library",
         "//pkg/util/pki:go_default_library",
+        "@io_k8s_api//core/v1:go_default_library",
         "@io_k8s_apimachinery//pkg/api/errors:go_default_library",
         "@io_k8s_client_go//listers/core/v1:go_default_library",
+        "@io_k8s_client_go//tools/record:go_default_library",
     ],
 )
 

--- a/pkg/controller/certificaterequests/selfsigned/selfsigned.go
+++ b/pkg/controller/certificaterequests/selfsigned/selfsigned.go
@@ -23,8 +23,10 @@ import (
 	"errors"
 	"fmt"
 
+	corev1 "k8s.io/api/core/v1"
 	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
 	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
 
 	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
 	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
@@ -40,6 +42,7 @@ import (
 
 const (
 	CRControllerName = "certificaterequests-issuer-selfsigned"
+	emptyDNMessage   = "Certificate will be issued with an empty Issuer DN, which contravenes RFC 5280 and could break some strict clients"
 )
 
 type signingFn func(*x509.Certificate, *x509.Certificate, crypto.PublicKey, interface{}) ([]byte, *x509.Certificate, error)
@@ -49,6 +52,7 @@ type SelfSigned struct {
 	secretsLister corelisters.SecretLister
 
 	reporter *crutil.Reporter
+	recorder record.EventRecorder
 
 	// Used for testing to get reproducible resulting certificates
 	signingFn signingFn
@@ -68,6 +72,7 @@ func NewSelfSigned(ctx *controllerpkg.Context) *SelfSigned {
 		issuerOptions: ctx.IssuerOptions,
 		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
 		reporter:      crutil.NewReporter(ctx.Clock, ctx.Recorder),
+		recorder:      ctx.Recorder,
 		signingFn:     pki.SignCertificate,
 	}
 }
@@ -126,6 +131,15 @@ func (s *SelfSigned) Sign(ctx context.Context, cr *cmapi.CertificateRequest, iss
 	}
 
 	template.CRLDistributionPoints = issuerObj.GetSpec().SelfSigned.CRLDistributionPoints
+
+	if template.Subject.String() == "" {
+		// RFC 5280 (https://tools.ietf.org/html/rfc5280#section-4.1.2.4) says that:
+		// "The issuer field MUST contain a non-empty distinguished name (DN)."
+		// Since we're creating a self-signed cert, the issuer will match whatever is
+		// in the template's subject DN.
+		log.V(logf.DebugLevel).Info("issued cert will have an empty issuer DN, which contravenes RFC 5280. emitting warning event")
+		s.recorder.Event(cr, corev1.EventTypeWarning, "BadConfig", emptyDNMessage)
+	}
 
 	// extract the public component of the key
 	publickey, err := pki.PublicKeyForPrivateKey(privatekey)

--- a/pkg/controller/certificaterequests/selfsigned/selfsigned_test.go
+++ b/pkg/controller/certificaterequests/selfsigned/selfsigned_test.go
@@ -25,6 +25,7 @@ import (
 	"encoding/asn1"
 	"encoding/pem"
 	"errors"
+	"fmt"
 	"testing"
 	"time"
 
@@ -51,9 +52,9 @@ var (
 	fixedClock      = fakeclock.NewFakeClock(fixedClockStart)
 )
 
-func generateCSR(t *testing.T, secretKey crypto.Signer, alg x509.SignatureAlgorithm) []byte {
+func generateCSR(t *testing.T, secretKey crypto.Signer, alg x509.SignatureAlgorithm, commonName string) []byte {
 	asn1Subj, _ := asn1.Marshal(pkix.Name{
-		CommonName: "test",
+		CommonName: commonName,
 	}.ToRDNSequence())
 	template := x509.CertificateRequest{
 		RawSubject:         asn1Subj,
@@ -106,7 +107,7 @@ func TestSign(t *testing.T) {
 			corev1.TLSPrivateKeyKey: []byte("this is a bad key"),
 		},
 	}
-	csrRSAPEM := generateCSR(t, skRSA, x509.SHA256WithRSA)
+	csrRSAPEM := generateCSR(t, skRSA, x509.SHA256WithRSA, "test-rsa")
 
 	skEC, err := pki.GenerateECPrivateKey(256)
 	if err != nil {
@@ -127,7 +128,9 @@ func TestSign(t *testing.T) {
 			corev1.TLSPrivateKeyKey: skECPEM,
 		},
 	}
-	csrECPEM := generateCSR(t, skEC, x509.ECDSAWithSHA256)
+	csrECPEM := generateCSR(t, skEC, x509.ECDSAWithSHA256, "test-ec")
+
+	csrEmptyCertPEM := generateCSR(t, skEC, x509.ECDSAWithSHA256, "")
 
 	baseCRNotApproved := gen.CertificateRequest("test-cr",
 		gen.SetCertificateRequestAnnotations(
@@ -163,6 +166,9 @@ func TestSign(t *testing.T) {
 	ecCR := gen.CertificateRequestFrom(baseCR,
 		gen.SetCertificateRequestCSR(csrECPEM),
 	)
+	emptyCR := gen.CertificateRequestFrom(baseCR,
+		gen.SetCertificateRequestCSR(csrEmptyCertPEM),
+	)
 
 	templateRSA, err := pki.GenerateTemplateFromCertificateRequest(baseCR)
 	if err != nil {
@@ -181,6 +187,18 @@ func TestSign(t *testing.T) {
 		t.FailNow()
 	}
 	certECPEM, _, err := pki.SignCertificate(templateEC, templateEC, skEC.Public(), skEC)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	templateEmptyCert, err := pki.GenerateTemplateFromCertificateRequest(emptyCR)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	emptyCertPEM, _, err := pki.SignCertificate(templateEmptyCert, templateEmptyCert, skEC.Public(), skEC)
 	if err != nil {
 		t.Error(err)
 		t.FailNow()
@@ -514,6 +532,50 @@ func TestSign(t *testing.T) {
 							}),
 							gen.SetCertificateRequestCertificate(certECPEM),
 							gen.SetCertificateRequestCA(certECPEM),
+						),
+					)),
+				},
+			},
+		},
+		"should sign a cert with no subject DN and create a warning event": {
+			certificateRequest: emptyCR.DeepCopy(),
+			signingFn: func(c1 *x509.Certificate, c2 *x509.Certificate, pk crypto.PublicKey, sk interface{}) ([]byte, *x509.Certificate, error) {
+				_, cert, err := pki.SignCertificate(c1, c2, pk, sk)
+				if err != nil {
+					return nil, nil, err
+				}
+
+				if cert.Subject.String() != "" {
+					return nil, nil, errors.New("invalid test: cert being issued should have an empty DN")
+				}
+
+				// need to return a known PEM cert as is done in other tets, since the actual issued cert (`cert` above)
+				// will have a different serial number + expiry
+				return emptyCertPEM, nil, nil
+			},
+			builder: &testpkg.Builder{
+				KubeObjects:        []runtime.Object{ecKeySecret},
+				CertManagerObjects: []runtime.Object{emptyCR.DeepCopy(), baseIssuer},
+				ExpectedEvents: []string{
+					fmt.Sprintf("Warning BadConfig %s", emptyDNMessage),
+					"Normal CertificateIssued Certificate fetched from issuer successfully",
+				},
+				ExpectedActions: []testpkg.Action{
+					testpkg.NewAction(coretesting.NewUpdateSubresourceAction(
+						cmapi.SchemeGroupVersion.WithResource("certificaterequests"),
+						"status",
+						gen.DefaultTestNamespace,
+						gen.CertificateRequestFrom(
+							emptyCR,
+							gen.SetCertificateRequestStatusCondition(cmapi.CertificateRequestCondition{
+								Type:               cmapi.CertificateRequestConditionReady,
+								Status:             cmmeta.ConditionTrue,
+								Reason:             cmapi.CertificateRequestReasonIssued,
+								Message:            "Certificate fetched from issuer successfully",
+								LastTransitionTime: &metaFixedClockStart,
+							}),
+							gen.SetCertificateRequestCertificate(emptyCertPEM),
+							gen.SetCertificateRequestCA(emptyCertPEM),
 						),
 					)),
 				},

--- a/pkg/controller/ingress-shim/sync.go
+++ b/pkg/controller/ingress-shim/sync.go
@@ -312,13 +312,6 @@ func setIssuerSpecificConfig(crt *cmapi.Certificate, ing *networkingv1beta1.Ingr
 	}
 }
 
-func setCommonName(crt *cmapi.Certificate, ing *networkingv1beta1.Ingress) {
-	// if annotation is set use that as CN
-	if ing.Annotations != nil && ing.Annotations[cmapi.CommonNameAnnotationKey] != "" {
-		crt.Spec.CommonName = ing.Annotations[cmapi.CommonNameAnnotationKey]
-	}
-}
-
 // shouldSync returns true if this ingress should have a Certificate resource
 // created for it
 func shouldSync(ing *networkingv1beta1.Ingress, autoCertificateAnnotations []string) bool {

--- a/pkg/util/pki/csr.go
+++ b/pkg/util/pki/csr.go
@@ -388,8 +388,8 @@ func GenerateTemplateFromCSRPEMWithUsages(csrPEM []byte, duration time.Duration,
 	}, nil
 }
 
-// SignCertificate returns a signed x509.Certificate object for the given
-// *v1.Certificate crt.
+// SignCertificate returns a signed *x509.Certificate given a template
+// *x509.Certificate crt and an issuer.
 // publicKey is the public key of the signee, and signerKey is the private
 // key of the signer.
 // It returns a PEM encoded copy of the Certificate as well as a *x509.Certificate

--- a/pkg/util/pki/generate.go
+++ b/pkg/util/pki/generate.go
@@ -38,11 +38,11 @@ const (
 	// generator functions in this package.
 	MaxRSAKeySize = 8192
 
-	// ECCurve256 represents a 256bit ECDSA key.
+	// ECCurve256 represents a secp256r1 / prime256v1 / NIST P-256 ECDSA key.
 	ECCurve256 = 256
-	// ECCurve384 represents a 384bit ECDSA key.
+	// ECCurve384 represents a secp384r1 / NIST P-384 ECDSA key.
 	ECCurve384 = 384
-	// ECCurve521 represents a 521bit ECDSA key.
+	// ECCurve521 represents a secp521r1 / NIST P-521 ECDSA key.
 	ECCurve521 = 521
 )
 

